File: MessageList/app/src/main/java/ru/startandroid/messagelist/base/mvp/BasePresenter.java
package ru.startandroid.messagelist.base.mvp;

import io.reactivex.disposables.Disposable;
import ru.startandroid.messagelist.common.utils.DisposableManager;

public abstract class BasePresenter<T extends MvpView> implements MvpPresenter<T> {

    private T view;
    private boolean firstCall = true;

    private final DisposableManager disposableManager = new DisposableManager();

    @Override
    public void attachView(T mvpView) {
        view = mvpView;
    }

    @Override
    public void detachView() {
        view = null;
    }

    @Override
    public void destroy() {
        disposableManager.clear();
    }

    public T getView() {
        return view;
    }

    protected boolean isViewAttached() {
        return view != null;
    }

    protected void addDisposable(Disposable disposable) {
        disposableManager.add(disposable);
    }

    protected void removeDisposable(Disposable disposable) {
            disposableManager.remove(disposable);
    }

    @Override
    public void viewIsReady() {
        onViewIsReady(firstCall);
        firstCall = false;
    }

    protected abstract void onViewIsReady(boolean firstCall);


}

File: Vocabulary/app/src/main/java/ru/startandroid/vocabulary/base/mvp/BasePresenter.kt
package ru.startandroid.vocabulary.base.mvp

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BasePresenter<T : MvpView> : MvpPresenter<T> {

    var view: T? = null
        private set

    var viewIsReadyFirstTime = true

    abstract fun onViewIsReady(firstTime: Boolean)

    override fun viewIsReady() {
        onViewIsReady(viewIsReadyFirstTime)
        viewIsReadyFirstTime = false

    }

    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    override fun attachView(mvpView: T) {
        view = mvpView
    }

    override fun detachView() {
        view = null
    }

    override fun destroy() {
        compositeDisposable.clear()
    }

    fun addDisposable(disposable: Disposable) {
        compositeDisposable.add(disposable)
    }

    fun removeDisposable(disposable: Disposable) {
        compositeDisposable.remove(disposable)
    }

}